### PR TITLE
Add timeline and countdown

### PR DESCRIPTION
**Pull request by 9r0x**
This commit adds timeline with countdown.

Closes https://github.com/coala/gci-leaders/issues/51

- On computer 
screen shot 2017-12-10 at 8 53 20 pm

- On mobile 
screen shot 2017-12-10 at 8 53 31 pm

### CODE REVIEW

File: lib/scrape.js [hunk 1]
```diff
@@ -105,3 +113,9 @@ async function fetchOrgsWithData() {
 
   fs.writeFileSync(`${__dirname}/../out/data.json`, JSON.stringify(data))
 })()
+
+
+;(async () => {
+  const dates = await fetchDates()
+  fs.writeFileSync(`${__dirname}/../out/dates.json`, JSON.stringify(dates))
```
Comment (blazeu) [inline lib/scrape.js]: Put this inside the IIFE block above.
No need to create a new one.

File: lib/scrape.js [hunk 2]
```diff
@@ -95,6 +95,14 @@ async function fetchOrgsWithData() {
   )
 }
 
+async function fetchDates() {
+  const res = await fetch(
+    'https://codein.withgoogle.com/api/program/current/'
+  )
+  const { result } = await res.json()
+  return result
```
Comment (blazeu) [inline lib/scrape.js]: This fails in CI, there's no `result` key in the json.
Just do `return res.json()`

File: lib/timeline.js [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+
+const noClaims = new Date(dates.competition_open_ends)
+
+const totalTime = (noClaims - competitionOpen)/(24*3600*1000)
+
+const timeLeft = (noClaims-datetime)/(24*3600*1000)
+
+const daysLeft = parseInt(timeLeft)
+
+const percentagePassed = parseInt(100-(timeLeft/totalTime)*100)
```
Comment (blazeu) [inline lib/timeline.js]: What's this file?

**Review by blazeu — CHANGES_REQUESTED** [review]
Remove .DS_Store and yarn.lock from the commit.

File: lib/generate.js [hunk 4]
```diff
@@ -2,11 +2,20 @@ const fs = require('fs')
...
+const today = new Date()
 const datetime = new Date(time).toUTCString()
 const rootURL = process.env.URL
 
+const competitionOpen = new Date(dates.competition_open_starts)
+const noClaims = new Date(dates.competition_open_ends)
+const totalTime = (noClaims - competitionOpen)/(24*3600*1000)
+const timeLeft = (noClaims-today)/(24*3600*1000)
+const daysLeft = parseInt(timeLeft)
+const percentagePassed = parseInt(100-(timeLeft/totalTime)*100)
```
Comment (blazeu) [inline lib/generate.js]: is `parseInt` even needed?

**Review by blazeu — CHANGES_REQUESTED** [review]

File: lib/generate.js (hunk 4 above)
Comment (9r0x) [reply lib/generate.js]: Otherwise it would have so many digits, so maybe not readable?

File: lib/timeline.js (hunk 3 above)
Comment (9r0x) [reply lib/timeline.js]: I tried to modularize the process, now it's removed

File: lib/generate.js [hunk 5]
```diff
@@ -2,11 +2,20 @@ const fs = require('fs')
...
+const today = new Date()
 const datetime = new Date(time).toUTCString()
 const rootURL = process.env.URL
 
+const competitionOpen = new Date(dates.competition_open_starts)
+const noClaims = new Date(dates.competition_open_ends)
+const totalTime = (noClaims - competitionOpen) / (24 * 3600 * 1000)
+const timeLeft = (noClaims - today) / (24 * 3600 * 1000)
+const daysLeft = parseInt(timeLeft)
+const percentagePassed = parseInt(100 - timeLeft / totalTime * 100)
```
Comment (blazeu) [inline lib/generate.js]: I prefer this to be in the frontend, because we might not do build everyday.

Use and modify `timeDifference` function inside `static/js/app.js`.

**Comment by blazeu** [comment]
Updated my comment there.

**Comment by 9r0x** [comment]
@blazeu  Sorry, I was having a class then, now I have repushed my code :P

File: static/js/app.js [hunk 6]
```diff
@@ -35,3 +35,31 @@ allLeaders.forEach(function(orgList) {
...
   }
 })
+
+document.getElementById('progress').textContent = gciProgress()
+
+function gciProgress() {
+  const competitionOpen = new Date(
+    document.getElementById('progress').dataset.competitionopen
+  )
+  console.log(competitionOpen)
```
Comment (blazeu) [inline static/js/app.js]: No `console.log` in frontend.

File: templates/main.html [hunk 7]
```diff
@@ -17,9 +17,18 @@
...
+        <div class="progress-inner">
+          <div class="progress-bar">
+          </div>
+        </div>
+      </div>
+      <span class="progress-text"><i id="progress" data-noclaims="{{noClaims}}" data-competitionopen="{{competitionOpen}}"></i></span>
+    </div>
     <hr>
     <i>
-      The leading participants for each organization are listed randomly.
```
Comment (blazeu) [inline templates/main.html]: Don't modify unrelated line.

File: .gitignore [hunk 8]
```diff
@@ -1,3 +1,5 @@
 node_modules
 out/*
 !out/.keep
+.DS_Store
+yarn.lock
```
Comment (blazeu) [inline .gitignore]: Another commit ? ;)

**Review by blazeu — CHANGES_REQUESTED** [review]

File: static/js/app.js [hunk 9]
```diff
@@ -35,3 +35,31 @@ allLeaders.forEach(function(orgList) {
...
+    .getElementsByClassName('progress-bar')[0]
+    .setAttribute(
+      'style',
+      'width:' + percentagePassed + '%;height: 8px;border-radius: 100px;'
+    )
+  return (
+    Math.round(percentagePassed) +
+    '% passed, ' +
+    Math.round(timeLeft) +
+    ' days before task claiming ends'
```
Comment (blazeu) [inline static/js/app.js]: Why not use `timeDifference` ?

File: static/js/app.js [hunk 10]
```diff
@@ -35,3 +35,31 @@ allLeaders.forEach(function(orgList) {
     orgList.appendChild(orgList.children[(Math.random() * i) | 0])
   }
 })
+
+document.getElementById('progress').textContent = gciProgress()
+
+function gciProgress() {
+  const competitionOpen = new Date(
+    document.getElementById('progress').dataset.competitionopen
```
Comment (blazeu) [inline static/js/app.js]: Storing the result of getElementById would be better instead of doing it again and again.

**Review by blazeu — APPROVED** [review]

**Comment by blazeu** [comment]
ack https://github.com/coala/gci-leaders/pull/60/commits/d4a895042c7e6f90c02dd04842eed05e63f43831

**Review by jayvdb — APPROVED** [review]
